src/labels.js

```diff
@@ -0,0 +1,443 @@
+var Metadata = require('./metadata');
+var Helpers = require('./helpers');
+var AddressHD = require('./address-hd');
+var BlockchainAPI = require('./api');
+
+var assert = require('assert');
+
+var METADATA_TYPE_LABELS = 4;
+
+class Labels {
+  constructor (metadata, wallet, object) {
+    this._readOnly = false; // Default
+    this._dirty = true;
+    this._wallet = wallet;
+    this._metadata = metadata;
+
+    let before = JSON.stringify(object);
+    object = this.migrateIfNeeded(object);
+
+    this.init(object);
+
+    if (JSON.stringify(object) !== before) {
+      this.save();
+    } else {
+      this._dirty = false;
+    }
+  }
+
+  get readOnly () {
+    return this._readOnly || !this._wallet.isMetadataReady;
+  }
+
+  get dirty () {
+    return this._dirty;
+  }
+
+  get version () {
+    return this._version;
+  }
+
+  init (object) {
+    this._version = object.version;
+    this._accounts = [];
+    for (let accountObject of object.accounts) {
+      let accountIndex = object.accounts.indexOf(accountObject);
+      let addresses = [];
+      for (let addressObject of accountObject) {
+        if (addressObject === null) {
+          addresses.push(null);
+        } else {
+          let hdAccount = this._wallet.hdwallet.accounts[accountIndex];
+          let addressIndex = accountObject.indexOf(addressObject);
+          addresses.push(new AddressHD(addressObject, hdAccount, addressIndex));
+        }
+      }
+      this._accounts[accountIndex] = addresses;
+    }
+  }
+
+  static initMetadata (wallet) {
+    return Metadata.fromMasterHDNode(wallet._metadataHDNode, METADATA_TYPE_LABELS);
+  }
+
+  toJSON () {
+    return {
+      version: this.version,
+      accounts: this._accounts.map((addresses) => {
+        if (addresses.length > 1) {
+          // Remove trailing null values:
+          while (!addresses[addresses.length - 1] || addresses[addresses.length - 1].label === null) {
+            addresses.pop();
```

Comment:
This will change the addresses array everywhere, intentional? `toJSON` generally doesn't have side effects.